feat: implement Toolbar and related components (DHIS2-19821)

edoardo opened the pull request:
Implements [DHIS2-19821](https://dhis2.atlassian.net/browse/DHIS2-19821)

### Description

Implements the `Toolbar` component with `UpdateButton`, `MenuBar` and `InterpretationsAndDetailsToggle` components.
The `MenuBar` component consists of `UpdateButton`, `FileMenu`, `ViewMenu`, `OptionsMenu` and `DownloadMenu` components.

---

### Quality checklist

Add _N/A_ to items that are not applicable and check them.

<!--Checkmate-->

-   [ ] Dashboard tested
-   [ ] Cypress and/or Jest tests added/updated
-   [ ] Docs added
-   [ ] d2-ci dependency replaced (requires <https://github.com/dhis2/analytics/pull/XXX>)
-   [ ] Tester approved (name)

---

### ToDos

<!--Checkmate-->

-   [ ] connect the components to the store

---

### Screenshots

<img width="1149" height="98" alt="Screenshot 2025-08-01 at 15 38 20" src="https://github.com/user-attachments/assets/8f6f489b-4c6b-4d2b-a2e5-48f2aa09b605" />


[DHIS2-19821]: https://dhis2.atlassian.net/browse/DHIS2-19821?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

edoardo committed:
chore: add history dependency

edoardo committed:
feat: add error utils and icons

edoardo committed:
feat: implement toolbar with menu bar and interpretations toggle

edoardo committed:
feat: add view menu in toolbar

edoardo committed:
feat: add download menu in toolbar

edoardo committed:
refactor: rename and move components for the menu bar

edoardo committed:
feat: add Options item in menu bar

HendrikThePendric commented on src/modules/visualization.ts:
```diff
@@ -0,0 +1,27 @@
+import { VIS_TYPE_LINE_LIST, VIS_TYPE_PIVOT_TABLE } from '@dhis2/analytics'
```
We should carefully consider, on a case-by-case basis to what extend the exported constants like this make sense in a TypeScript project. I think in many cases (this is a good example) they actually don't. Instead we should be using string [literal types](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#literal-types).

So here we should probably export a type called `VisualizationTypes` from `src/types/analytics`, which includes all possible types. And then in our project we should probably have a type like this
```typescript
type EventVisualisationTypes = Extract<VisualizationTypes, 'LINE_LIST' | 'PIVOT_TABLE'>
```

I also just realised that perhaps we should not have `VisualizationTypes` in analytics at all. The generated types has an enum called `EventVisualizationType`. So we should probably have a type like this in `src/types/index.ts`:
```
type SupportedVisTypes = Extract<
    (typeof EventVisualizationType)[keyof typeof EventVisualizationType],
    | 'LINE_LIST'
    | 'PIVOT_TABLE'
>
```

HendrikThePendric reviewed (commented):
I've left a few comments in the code, and would also like to leave a general note...

I think it is good you have started porting components from DV/LL to the new project this way. You have taken everything from the old project and commented out what doesn't work. I am not completely sure how I would have gone about this, or what the best approach is. I guess this works well, and it will prevent us from re-inventing the wheel.

But on the other hand I do fear that it could potentially lead us to take problematic parts from our old apps to the new one. I suppose if we take time to review carefully and ensure that we don't make the PRs too big, it is probably the most progmatic way of working.

edoardo committed:
chore: type icon components

edoardo commented on src/components/download-menu/use-download.ts:
```diff
@@ -0,0 +1,205 @@
+//    }
+//}
+
+const useDownload = () => ({
+    isDownloadDisabled: false,
+    download: () => console.log('download fn TBD'),
+})
```
Can then the download function skip the types in the arguments?

```suggestion
const useDownload = (): UseDownloadResult => ({
    isDownloadDisabled: false,
    download: (type, format, idScheme) => {
      console.log(`TBD (type: ${type}, format: ${format}, idScheme: ${idScheme})`),
})
```

edoardo committed:
refactor: add types for download menu (type, format, idScheme)

edoardo committed:
chore: add type for React component

edoardo committed:
refactor: prefer named exports

edoardo committed:
chore: add types for analytics components

edoardo committed:
refactor: use string literal instead of constant from analytics

edoardo committed:
fix: avoid "possibly undefined" type error

serverVersion might be undefined and in that case the numeric comparison
throws an error too.
Avoid reading from undefined error and use a default (0) in the
comparison when serverVersion is undefined.

edoardo commented on src/components/download-menu/download-menu.tsx:
```diff
@@ -0,0 +1,78 @@
+import PlainDataSourceSubMenu from './plain-data-source-sub-menu'
+
+type DownloadMenuProps = {
+    download: (type: string, format: string, idScheme?: string) => void
```
Fixed.

edoardo committed:
chore: fix imports

edoardo committed:
feat: add current store slice

Fixed imports and tests.

edoardo committed:
feat: add constants for the panels

edoardo committed:
feat: add module for localStorage getters/setters

edoardo committed:
fix: read accessory sidebar width from user preferences

edoardo committed:
feat: connect interpretations/details toggler to the store

edoardo committed:
feat: store accessory sidebar width as user preference

edoardo committed:
fix: fix imports from constants and modules

edoardo commented on src/store/ui-slice.ts:
```diff
@@ -1,27 +1,85 @@
 export const initialState: UiState = {
     visualizationType: 'LINE_LIST',
+
+    hideLayoutPanel: false,
```
These are copied from LL.
I wonder about the naming, there's a mix of `hide` and `show` here, I assume the reason is the default state, if the panel is showing by default, the key/action is `hide` and viceversa.

edoardo committed:
feat: add visualization slice and use generated types for it and current

edoardo committed:
feat: add type for eventVisualization object

edoardo committed:
feat: export setVisualization from visualization slice

edoardo committed:
chore: add return types for local storage functions

edoardo committed:
feat: implement delete with rtk mutation

jenniferarnesen commented:
Superseded
